### PR TITLE
feat(connector): Add support for complete authorize payment after 3DS redirection

### PR DESCRIPTION
**Pull request by jagan-jaya**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [X] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description

1. For Connectors like Nuvei, we need to call a complete payment call after the 3DS authnetication is completed. this should call the authorize API with reference id from previous 3ds call. If the authentication is successful the liability shift will move to issuer, in case of failure there won't be any trace of this payment in the connector (Psync won't work).  Adding `Complete Authorize` flow to support the functionality
2. Move default trait implementation for connectors to macros, so we can make the default implementation in one file instead of each connector file. Credits: @jarnura

### Additional Changes

- [ ] This PR modifies the database schema
- [X] This PR modifies application configuration/environment variables

## Motivation and Context

Add support for complete authorize flow after redirection (3DS, paypal)

## How did you test it?

Screen Shot 2023-03-14 at 3 45 21 AM
Screen Shot 2023-03-14 at 3 46 07 AM

## Checklist

- [X] I formatted the code `cargo +nightly fmt --all`
- [X] I addressed lints thrown by `cargo clippy`
- [X] I reviewed submitted code

### CODE REVIEW

File: crates/router/src/core/payments/flows/authorize_flow.rs [hunk 1]
```diff
@@ -118,6 +119,14 @@ impl types::PaymentsAuthorizeRouterData {
             _ => Ok(self.clone()),
         }
     }
+
+    fn decide_authentication_type(&mut self) {
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/flows/authorize_flow.rs:123]: if the merchant passes 3ds auth then how can we change it to no_3ds

File: crates/router/src/core/payments/operations/payment_complete_authorize.rs [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+        db: &dyn StorageInterface,
+        _payment_id: &api::PaymentIdType,
+        mut payment_data: PaymentData<F>,
+        customer: Option<storage::Customer>,
+        storage_scheme: storage_enums::MerchantStorageScheme,
+    ) -> RouterResult<(BoxedOperation<'b, F, api::PaymentsRequest>, PaymentData<F>)>
+    where
+        F: 'b + Send,
+    {
+        let payment_method = payment_data.payment_attempt.payment_method;
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/operations/payment_complete_authorize.rs:298]: There should not be any uupdates happening in this operation, only post update tracker should be there right?

File: crates/router/src/core/payments/flows/authorize_flow.rs (hunk 1 above)
Comment (jagan-jaya) [reply crates/router/src/core/payments/flows/authorize_flow.rs:123]: let say the merchant sends 3DS, and when we are checking the card enrolled we see that the card is not enrolled for 3DS v2, in that case we decide to go with non-3ds flow. we can make this configurable in merchant_connector_meta

File: crates/router/src/connector/utils.rs [hunk 3]
```diff
@@ -352,6 +370,102 @@ pub fn get_header_key_value<'a>(
...
+
+pub fn parse_struct<T>(json: serde_json::Value) -> Result<T, Error>
+where
+    T: serde::de::DeserializeOwned,
+{
+    serde_json::from_value::<T>(json.clone())
+        .into_report()
+        .change_context(errors::ConnectorError::ParsingFailed {
+            from_type: "Json",
+            to_type: std::any::type_name::<T>(),
```
Comment (jarnura) [inline crates/router/src/connector/utils.rs:418]: `type_name` this is a unstable function and the result is unpredictable

File: crates/router/src/connector/utils.rs [hunk 4]
```diff
@@ -352,6 +370,102 @@ pub fn get_header_key_value<'a>(
...
+where
+    T: serde::de::DeserializeOwned,
+{
+    let connector_meta_secret =
+        connector_meta.ok_or_else(missing_field_err("connector_meta_data"))?;
+    let json = connector_meta_secret.peek().clone();
+    parse_struct(json)
+}
+
+pub fn parse_struct<T>(json: serde_json::Value) -> Result<T, Error>
```
Comment (jarnura) [inline crates/router/src/connector/utils.rs:410]: already there is a util function, why we need a another one?

File: crates/router/src/connector/utils.rs [hunk 5]
```diff
@@ -352,6 +370,102 @@ pub fn get_header_key_value<'a>(
...
+                .into_report()
+                .change_context(errors::ConnectorError::ResponseHandlingFailed)?,
+        })
+}
+
+pub fn to_string<T>(data: &T) -> Result<String, Error>
+where
+    T: serde::Serialize,
+{
+    serde_json::to_string(data)
```
Comment (jarnura) [inline crates/router/src/connector/utils.rs:457]: use the common util function from ` crates/common_utils/src/ext_traits.rs`.

File: crates/router/src/connector/utils.rs [hunk 6]
```diff
@@ -352,6 +370,102 @@ pub fn get_header_key_value<'a>(
...
+                .into_report()
+                .change_context(errors::ConnectorError::ResponseHandlingFailed)?,
+        })
+}
+
+pub fn parse_struct_from_bytes_slice<T>(bytes: &[u8]) -> Result<T, Error>
+where
+    T: serde::de::DeserializeOwned,
+{
+    serde_json::from_slice::<T>(bytes)
```
Comment (jarnura) [inline crates/router/src/connector/utils.rs:442]: there is crate which has common utils function, check this file ` crates/common_utils/src/ext_traits.rs`

File: crates/router/src/connector/utils.rs [hunk 7]
```diff
@@ -352,6 +370,102 @@ pub fn get_header_key_value<'a>(
...
+
+pub fn to_connector_meta<T>(connector_meta: Option<serde_json::Value>) -> Result<T, Error>
+where
+    T: serde::de::DeserializeOwned,
+{
+    let json = connector_meta.ok_or_else(missing_field_err("connector_meta_data"))?;
+    parse_struct(json)
+}
+
+pub fn to_connector_meta_from_secret<T>(
```
Comment (jarnura) [inline crates/router/src/connector/utils.rs:398]: Try to implement the function as chained function. Can be done in a separate PR too.

File: crates/router/src/core/payments.rs [hunk 8]
```diff
@@ -18,7 +18,7 @@ pub use self::operations::{
 };
 use self::{
     flows::{ConstructFlowSpecificData, Feature},
-    operations::{BoxedOperation, Operation},
+    operations::{payment_complete_authorize::CompleteAuthorize, BoxedOperation, Operation},
```
Comment (jarnura) [inline crates/router/src/core/payments.rs:21]: remove specific imports

File: crates/router/src/core/payments.rs [hunk 9]
```diff
@@ -29,7 +29,7 @@ use crate::{
     logger, pii,
     routes::AppState,
     scheduler::utils as pt_utils,
-    services,
+    services::{self, PaymentAction},
```
Comment (jarnura) [inline crates/router/src/core/payments.rs:32]: remove specific imports

File: crates/router/src/core/payments.rs [hunk 10]
```diff
@@ -193,7 +193,11 @@ where
                 .await?
             }
         };
-        vault::Vault::delete_locker_payment_method_by_lookup_key(state, &payment_data.token).await
+        if payment_data.payment_intent.status != storage_enums::IntentStatus::RequiresCustomerAction
```
Comment (jarnura) [inline crates/router/src/core/payments.rs:196]: Why this change is added?

**Review by jarnura — COMMENTED** [review]
From next time on wards give micro pull request.

File: crates/router/src/connector/nuvei/transformers.rs [hunk 11]
```diff
@@ -229,53 +346,230 @@ impl<F, T>
...
+                        ..Default::default()
+                    },
+                    billing_address: Some(BillingAddress {
+                        email: item.request.get_email()?,
+                        country: item.get_billing_country()?,
+                    }),
+                    ..Default::default()
+                }),
+                _ => Err(
+                    errors::ConnectorError::NotImplemented("Payment methods".to_string()).into(),
```
Comment (Narayanbhat166) [inline crates/router/src/connector/nuvei/transformers.rs:419]: There is a `NotSupported` connector error, I think we can use that if this connector does not support a particular payment method

File: crates/router/src/connector/nuvei/transformers.rs [hunk 12]
```diff
@@ -229,53 +346,230 @@ impl<F, T>
...
+        })
+    } else {
+        None
+    };
+    let card = card_details.clone();
+    Ok(NuveiPaymentsRequest {
+        related_transaction_id,
+        payment_option: PaymentOption {
+            card: Some(Card {
+                card_number: Some(card.card_number),
```
Comment (Narayanbhat166) [inline crates/router/src/connector/nuvei/transformers.rs:488]: This should not be optional since it is an enum, the fields should be mandatory

File: crates/router/src/connector/nuvei/transformers.rs [hunk 13]
```diff
@@ -521,10 +822,28 @@ impl<F, T>
...
+            .payment_option
+            .as_ref()
+            .and_then(|o| o.card.clone())
+            .and_then(|card| card.three_d)
+            .and_then(|three_ds| {
+                if let (Some(acs_url), Some(c_req)) = (three_ds.acs_url, three_ds.c_req) {
+                    Some((acs_url, c_req))
+                } else {
+                    None
+                }
```
Comment (Narayanbhat166) [inline crates/router/src/connector/nuvei/transformers.rs:840]:
```suggestion
three_ds.acs_url.zip(three_ds.c_req)
```

File: crates/router/src/connector/utils.rs [hunk 14]
```diff
@@ -352,6 +370,102 @@ pub fn get_header_key_value<'a>(
         ))?
 }
 
+pub fn to_boolean(string: String) -> bool {
```
Comment (Narayanbhat166) [inline crates/router/src/connector/utils.rs:373]: what is the requirement for this? I see this function is not being used anywhere, can be removed

File: crates/router/src/connector/utils.rs [hunk 15]
```diff
@@ -352,6 +370,102 @@ pub fn get_header_key_value<'a>(
...
+    T: serde::de::DeserializeOwned,
+{
+    serde_json::from_value::<T>(json.clone())
+        .into_report()
+        .change_context(errors::ConnectorError::ParsingFailed {
+            from_type: "Json",
+            to_type: std::any::type_name::<T>(),
+            data: json.to_string(),
+        })
+}
```
Comment (Narayanbhat166) [inline crates/router/src/connector/utils.rs:421]: these functions already exist in `common_utils`

**Review by Narayanbhat166 — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/utils.rs (hunk 4 above)
Comment (jagan-jaya) [reply crates/router/src/connector/utils.rs:410]: we need this to handle connector error, as we duplication connector error handling in all the places it will solve the problem

File: crates/router/src/connector/utils.rs (hunk 5 above)
Comment (jagan-jaya) [reply crates/router/src/connector/utils.rs:457]: we need this to handle connector error, as we duplication connector error handling in all the places it will solve the problem

File: crates/router/src/connector/utils.rs (hunk 6 above)
Comment (jagan-jaya) [reply crates/router/src/connector/utils.rs:442]: we need this to handle connector error, as we duplication connector error handling in all the places it will solve the problem

File: crates/router/src/core/payments.rs (hunk 10 above)
Comment (jagan-jaya) [reply crates/router/src/core/payments.rs:196]: We need to keep the token till will get the terminal status, RequiresCustomerAction is not a terminal status so it should be deleted only after completing authorize step

File: crates/router/src/connector/utils.rs (hunk 14 above)
Comment (jagan-jaya) [reply crates/router/src/connector/utils.rs:373]: it is being used in nuvei.rs::695

File: crates/router/src/connector/utils.rs (hunk 4 above)
Comment (jarnura) [reply crates/router/src/connector/utils.rs:410]: Refer `crates/common_utils/src/errors.rs`, there is trait called `ErrorSwitch<T>` which needs to implemented as `ErrorSwitch<ConnectorError> for ParsingError`. Then all utils functions with `ParsingError` can be used inside `ConnectorError` context also.

Eg: json.parse_struct("typename").switch()

**Review by jarnura — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]
